File: src/simulator/robot/sensors/Laser.java
package simulator.robot.sensors;

import math.Vec2;
import processing.core.PApplet;
import simulator.Simulator;
import simulator.environment.Landmark;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

public class Laser {
    public static final int COUNT = 181;

    // Angle
    public static final double MIN_THETA = -Math.PI / 2;
    public static final double MAX_THETA = Math.PI / 2;
    public static final double ANGULAR_RESOLUTION = (MAX_THETA - MIN_THETA) / COUNT;
    public static final double ANGLE_ERROR_LIMIT = 0.05;

    // Distance
    public static final double MAX_DISTANCE = 5.0;
    public static final double INVALID_MEASUREMENT_VALUE = MAX_DISTANCE + 1;
    public static final double DISTANCE_ERROR_LIMIT = 0.05;

    // Graphics
    private final PApplet applet;

    // Multi-thread access
    private final List<Double> measurements = new ArrayList<>(COUNT);

    public Laser(PApplet applet) {
        this.applet = applet;
        for (int i = 0; i < COUNT; i++) {
            measurements.add(INVALID_MEASUREMENT_VALUE);
        }
    }

    public void updateLaserScan(Vec2 position, double orientation, List<Landmark> landmarks) {
        List<Double> newMeasurements = new ArrayList<>(Laser.COUNT);
        for (int i = 0; i < Laser.COUNT; i++) {
            newMeasurements.add(Laser.INVALID_MEASUREMENT_VALUE);
        }

        // For each laser beam
        for (int i = 0; i < Laser.COUNT; ++i) {
            double percentage = i / (Laser.COUNT - 1.0);
            double laserThErr = ThreadLocalRandom.current().nextDouble(-Laser.ANGLE_ERROR_LIMIT * Laser.ANGULAR_RESOLUTION, Laser.ANGLE_ERROR_LIMIT * Laser.ANGULAR_RESOLUTION);
            double theta = Laser.MIN_THETA + (Laser.MAX_THETA - Laser.MIN_THETA) * percentage + orientation + laserThErr;
            Vec2 v = Vec2.of(Math.cos(theta), Math.sin(theta));

            // Check intersection for each line feature
            for (Landmark landmark : landmarks) {
                double rayDistance = landmark.shortestRayDistanceFrom(position, v);
                if (rayDistance >= 0 && rayDistance < Laser.MAX_DISTANCE) {
                    newMeasurements.set(i, Math.min(rayDistance, newMeasurements.get(i)));
                }
            }

            // Add some noise to new measurements
            if (newMeasurements.get(i) < Laser.INVALID_MEASUREMENT_VALUE) {
                double laser_d_err = ThreadLocalRandom.current().nextDouble(-Laser.DISTANCE_ERROR_LIMIT, Laser.DISTANCE_ERROR_LIMIT);
                newMeasurements.set(i, newMeasurements.get(i) + laser_d_err);
            }
        }

        // Update measurements
        synchronized (measurements) {
            for (int i = 0; i < newMeasurements.size(); i++) {
                measurements.set(i, newMeasurements.get(i));
            }
        }
    }

    public List<Double> getMeasurements() {
        List<Double> currentMeasurements;
        synchronized (measurements) {
            currentMeasurements = new ArrayList<>(measurements);
        }
        return currentMeasurements;
    }

    public void draw(Vec2 position, double orientation) {
        List<Double> distances = getMeasurements();
        List<Vec2> lasers = new ArrayList<>(Laser.COUNT);
        for (int i = 0; i < distances.size(); ++i) {
            if (distances.get(i) == Laser.INVALID_MEASUREMENT_VALUE) {
                continue;
            }
            double percentage = i / (Laser.COUNT - 1.0);
            double theta = Laser.MIN_THETA + (Laser.MAX_THETA - Laser.MIN_THETA) * percentage;

            Vec2 laserSprite = position.plus(Vec2.of(Math.cos(orientation + theta), Math.sin(orientation + theta)).scaleInPlace(distances.get(i) * Simulator.SCALE));
            lasers.add(laserSprite);
        }
        applet.stroke(1, 0, 0);
        for (Vec2 l : lasers) {
            applet.line((float) position.x, (float) position.y, (float) l.x, (float) l.y);
        }
    }
}

File: src/simulator/Laser.kt
package simulator

import extensions.plus
import extensions.timesAssign
import org.ejml.data.DMatrix2
import processing.core.PApplet
import java.util.*
import java.util.concurrent.ThreadLocalRandom
import kotlin.math.cos
import kotlin.math.sin

class Laser internal constructor(private val applet: PApplet) {

    companion object {
        const val COUNT = 181

        // Angle
        const val MIN_THETA = -Math.PI / 2
        const val MAX_THETA = Math.PI / 2
        const val ANGULAR_RESOLUTION = (MAX_THETA - MIN_THETA) / COUNT
        var ANGLE_ERROR_LIMIT = 0.05

        // Distance
        var MAX_DISTANCE = 500.0
        var DISTANCE_ERROR_LIMIT = 0.05
        private fun invalidMeasurementValue(): Double {
            return MAX_DISTANCE + 1
        }
    }

    // Multi-thread access
    private val measurements: MutableList<Double> = ArrayList(COUNT)

    init {
        for (i in 0 until COUNT) {
            measurements.add(invalidMeasurementValue())
        }
    }

    fun updateLaserScan(position: DMatrix2, orientation: Double, landmarks: List<Landmark>) {
        val newMeasurements: MutableList<Double> = ArrayList(COUNT)
        for (i in 0 until COUNT) {
            newMeasurements.add(invalidMeasurementValue())
        }

        // For each laser beam
        for (i in 0 until COUNT) {
            val percentage = i / (COUNT - 1.0)
            val laserThErr = ThreadLocalRandom.current().nextDouble(-ANGLE_ERROR_LIMIT * ANGULAR_RESOLUTION, ANGLE_ERROR_LIMIT * ANGULAR_RESOLUTION)
            val theta = MIN_THETA + (MAX_THETA - MIN_THETA) * percentage + orientation + laserThErr
            val v = DMatrix2(cos(theta), sin(theta))

            // Check intersection for each line feature
            for (landmark in landmarks) {
                val rayDistance = landmark.shortestRayDistanceFrom(position, v)
                if (rayDistance >= 0 && rayDistance < MAX_DISTANCE) {
                    newMeasurements[i] = Math.min(rayDistance, newMeasurements[i])
                }
            }

            // Add some noise to new measurements
            if (newMeasurements[i] < invalidMeasurementValue()) {
                val laser_d_err = ThreadLocalRandom.current().nextDouble(-DISTANCE_ERROR_LIMIT, DISTANCE_ERROR_LIMIT)
                newMeasurements[i] = newMeasurements[i] + laser_d_err
            }
        }

        // Update measurements
        synchronized(measurements) {
            for (i in newMeasurements.indices) {
                measurements[i] = newMeasurements[i]
            }
        }
    }

    fun getMeasurements(): List<Double> {
        var currentMeasurements: List<Double>
        synchronized(measurements) { currentMeasurements = ArrayList(measurements) }
        return currentMeasurements
    }

    fun draw(position: DMatrix2, orientation: Double) {
        val distances = getMeasurements()
        val lasers: MutableList<DMatrix2> = ArrayList(COUNT)
        for (i in distances.indices) {
            if (distances[i] == invalidMeasurementValue()) {
                continue
            }
            val percentage = i / (COUNT - 1.0)
            val theta = MIN_THETA + (MAX_THETA - MIN_THETA) * percentage
            val laserBeam = DMatrix2(cos(orientation + theta),
                    sin(orientation + theta))
            laserBeam *= distances[i]
            val laserEnd = position + laserBeam
            lasers.add(laserEnd)
        }
        applet.stroke(1f, 0f, 0f)
        for (l in lasers) {
            applet.line(position.a1.toFloat(), 0f, position.a2.toFloat(), l.a1.toFloat(), 0f, l.a2.toFloat())
        }
    }

}